### DISCUSSION TITLE
if you would like to provide a docker version？

### DISCUSSION THREAD

**Discussion by liangaomng** [dynamicslab/hydrogym#108] [Q&A]
Hello，
Prof.Steve, It seems that the firedrake has some problems with installing the environment. 
I wonder if your group would like to provide a docker version that includes simulation software？

best,
aoming

**Comment by ludgerpaehler** [comment]
Dear Aoming,

We are currently doing a large refactoring to land a new version before APS. The new version will fix the issues :) 

Best,
Ludger

**Comment by liangaomng** [comment]
thx!

**Comment by SamAhnert** [comment]
Hi Aoming,

Can I ask what issues you are currently having? I know Ludger has some things in the works, but I do want to clarify that as of last week I have built a new version of firedrake and successfully installed HydroGym using pip. The tutorials aren't working since the data files were generated under the old version of firedrake checkpointing (I just found this out yesterday), but if you go to the test folder and run `pytest test_cyl.py`, for example, all test should pass. You should be able to interact with the solver or environment accordingly (different tests in our suite interact with HydroGym at different levels) if you want to familiarize yourself with the API.

**Comment by liangaomng** [comment]
> Hi Aoming,
> 
> Can I ask what issues you are currently having? I know Ludger has some things in the works, but I do want to clarify that as of last week I have built a new version of firedrake and successfully installed HydroGym using pip. The tutorials aren't working since the data files were generated under the old version of firedrake checkpointing (I just found this out yesterday), but if you go to the test folder and run `pytest test_cyl.py`, for example, all test should pass. You should be able to interact with the solver or environment accordingly (different tests in our suite interact with HydroGym at different levels) if you want to familiarize yourself with the API.

Hello Sam,

Yes, I encounter an error when I install Firedfrake in the accordingDOCS:"python3 firedrake-install --venv-name=path/to/venv". And it shows that  
{" File "/cluster/home/xuzhiqin_10/lam/hydrogym/third_party/firedrake/scripts/firedrake-install", line 1675, in <module>
    check_call(["apt-get", "--version"])
  File "/cluster/home/xuzhiqin_10/lam/hydrogym/third_party/firedrake/scripts/firedrake-install", line 653, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
  File "/cluster/home/xuzhiqin_10/lam/miniconda3/envs/myenv/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/cluster/home/xuzhiqin_10/lam/miniconda3/envs/myenv/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/cluster/home/xuzhiqin_10/lam/miniconda3/envs/myenv/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/cluster/home/xuzhiqin_10/lam/miniconda3/envs/myenv/lib/python3.10/subprocess.py", line 1863, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: 'apt-get'}. And I add sudo to exec it. It shows {that my account is not in the sudoers file.  This incident will be reported. } Do you know how to fix？
Best，
AOMING

**Comment by liangaomng** [comment]
> day), but if you go to the test folder and run `pytest test_cyl.py`, for example, all test should pass. You should be able to interact with the solver or environment accordingly (diff

rootdir: /cluster/home/xuzhiqin_10/lam/hydrogym/test
collected 0 items / 1 error                                                                                                                                                         

====================================================================================== ERRORS =======================================================================================
___________________________________________________________________________ ERROR collecting test_cyl.py ____________________________________________________________________________
ImportError while importing test module '/cluster/home/xuzhiqin_10/lam/hydrogym/test/test_cyl.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../miniconda3/envs/myenv/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_cyl.py:3: in <module>
    import firedrake as fd
E   ModuleNotFoundError: No module named 'firedrake'

**Comment by liangaomng** [comment]
Hi Prof，

I choose other computer，but it shows thatCreating firedrake venv in '/liangaoming/hydrogym/third_party/firedrake/scripts/path/to/venv'.
Traceback (most recent call last):
  File "firedrake-install", line 1714, in <module>
    run_pip(["install", "-U", "setuptools==59.2.0"])  # Unpin this when numpy will build with latest setuptools
  File "firedrake-install", line 956, in run_pip
    check_call(pip + args)
  File "firedrake-install", line 653, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
  File "/usr/lib/python3.6/subprocess.py", line 356, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/liangaoming/hydrogym/third_party/firedrake/scripts/path/to/venv/bin/python', '-m', 'pip', 'install', '-U', 'setuptools==59.2.0']' returned non-zero exit status 2.

**Comment by liangaomng** [comment]
update：
Eigen tarball hash valid
Depending on your platform, PETSc may take an hour or more to build!
Installing pip dependencies for loopy
Installing pip dependencies for petsc4py
No dependencies found. Skipping.
Installing pip dependencies for ufl
No dependencies found. Skipping.
Installing pip dependencies for fiat
No dependencies found. Skipping.
Installing pip dependencies for FInAT
Traceback (most recent call last):
  File "/liangaoming/hydrogym/third_party/firedrake/scripts/firedrake-install", line 1864, in <module>
    pip_requirements(p, compiler_env)
  File "/liangaoming/hydrogym/third_party/firedrake/scripts/firedrake-install", line 991, in pip_requirements
    run_pip_install_wrap(reqs, parallel_compiler_env)
  File "/liangaoming/hydrogym/third_party/firedrake/scripts/firedrake-install", line 984, in run_pip_install_wrap
    run_pip_install([req])
  File "/liangaoming/hydrogym/third_party/firedrake/scripts/firedrake-install", line 965, in run_pip_install
    check_call(pipinstall + pipargs)
  File "/liangaoming/hydrogym/third_party/firedrake/scripts/firedrake-install", line 653, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
  File "/root/liangaoming/miniconda3/envs/myenv/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/root/liangaoming/miniconda3/envs/myenv/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/liangaoming/hydrogym/third_party/firedrake/scripts/venv/bin/python', '-m', 'pip', 'install', '--no-build-isolation', '--no-binary', 'mpi4py,randomgen,islpy,numpy', '-vvv', 'numpy>=1.16']' returned non-zero exit status 2.

Do you know how to fix？
Aoming

**Comment by SamAhnert** [comment]
Ah okay, if I read this correctly you're trying to install firedrake from a python binary file that was generated using miniconda. Firedrake is not compatible with conda environments at all. You can see the "System anti-requirements" sections of their [installation guide](https://www.firedrakeproject.org/download.html) for more details. They recommend removing conda from your system, but if you are actively using conda environments (which I think is a fair and very common practice nowadays) you should be able to download a docker image of firedrake that will be separate from your system. I haven't tested this in a while, but if you successfully build the docker image, you should be able to pip install hydrogym in that image. 

We recognize that this is a quite large hurdle to begin using firedrake/HydroGym, and perhaps another option could be to create a new user on your desktop that you use only for firedrake. I haven't tried this before, but myself and my collaborators don't see any reason why this wouldn't work so long as you don't install conda on that user. If you choose to go down this route, I recommend installing firedrake with Python3.10 since firedrake will install for python 3.8-3.11, but Hydrogym requires 3.10 in it's current state.

If you do keep trying let me know how it goes! We hope to lower the barrier to entry soon with our next major update!

Best,
Sam

**Reply by ReHoss** [reply]
Installing Firedrake is the hard part of the process!
